dex/networks/ltc/tx.go

```diff
@@ -0,0 +1,556 @@
+		// proof RangeProof
+		// signature
+		if err = d.discardBytes(33 + 33 + 33); err != nil { // commitment, sender pk, receiver pk
+			return nil, err
+		}
+
+		// OutputMessage
+		feats, err := d.readByte()
+		if err != nil {
+			return nil, err
+		}
+		// enum FeatureBit {
+		// 	STANDARD_FIELDS_FEATURE_BIT = 0x01,
+		// 	EXTRA_DATA_FEATURE_BIT = 0x02
+		// };
+		if feats&0x1 != 0 { // pubkey | view_tag uint8_t | masked_value uint64_t | nonce 16-bytes
+			if err = d.discardBytes(33 + 1 + 8 + 16); err != nil {
+				return nil, err
+			}
+		}
+		if feats&0x2 != 0 { // extraData
+			if err = d.discardVect(); err != nil {
+				return nil, err
+			}
+		}
+
+		// RangeProof "The proof itself, at most 675 bytes long."
+		// std::vector<uint8_t> m_bytes; -- except it's actually used like a [675]byte...
+		if err = d.discardBytes(675 + 64); err != nil { // proof + sig
+			return nil, err
+		}
+	} // outputs
+
+	// kernels
+	numKerns, err := d.readCompactSize()
+	if err != nil {
+		return nil, err
+	}
+	// Capture the first kernel since pure MW txns (no canonical inputs or
+	// outputs) that live in the MWEB but are also seen in mempool have their
+	// hash computed as blake3_256(kernel0).
+	var kern0 []byte
+	d.tee = new(bytes.Buffer)
+	for i := 0; i < int(numKerns); i++ {
+		// src/libmw/include/mw/models/tx/Kernel.h - class Kernel
+
+		// enum FeatureBit {
+		//     FEE_FEATURE_BIT = 0x01,
+		//     PEGIN_FEATURE_BIT = 0x02,
+		//     PEGOUT_FEATURE_BIT = 0x04,
+		//     HEIGHT_LOCK_FEATURE_BIT = 0x08,
+		//     STEALTH_EXCESS_FEATURE_BIT = 0x10,
+		//     EXTRA_DATA_FEATURE_BIT = 0x20,
+		//     ALL_FEATURE_BITS = FEE_FEATURE_BIT | PEGIN...
+		// };
+		feats, err := d.readByte()
+		if err != nil {
+			return nil, err
+		}
+		if feats&0x1 != 0 { // fee
+			_, err = d.readVLQ() // vlq for amount? in the wire protocol?!?
+			if err != nil {
+				return nil, err
+			}
+		}
+		if feats&0x2 != 0 { // pegin amt
+			_, err = d.readVLQ()
+			if err != nil {
+				return nil, err
+			}
+		}
+		if feats&0x4 != 0 { // pegouts vector
+			sz, err := d.readCompactSize()
+			if err != nil {
+				return nil, err
+			}
+			for i := uint64(0); i < sz; i++ {
+				_, err = d.readVLQ() // pegout amt
+				if err != nil {
+					return nil, err
+				}
+				if err = d.discardVect(); err != nil { // pkScript
+					return nil, err
+				}
+			}
+		}
+		if feats&0x8 != 0 { // lockHeight
+			_, err = d.readVLQ()
+			if err != nil {
+				return nil, err
+			}
+		}
+		if feats&0x10 != 0 { // stealth_excess pubkey
+			if err = d.discardBytes(33); err != nil {
+				return nil, err
+			}
+		}
+		if feats&0x20 != 0 { // extraData vector
+			if err = d.discardVect(); err != nil {
+				return nil, err
+			}
+		}
+		// "excess" commitment and signature
+		if err = d.discardBytes(33 + 64); err != nil {
+			return nil, err
+		}
+
+		if i == 0 {
+			kern0 = d.tee.Bytes()
+			d.tee = nil
+		}
+	} // kernels
+
+	return kern0, nil
+}
+
+type Tx struct {
+	*wire.MsgTx
+	IsHogEx bool
```

Comment:
Thanks god they shortened it to "HogEx". I don't know if I could take it seriously if it said **Hogwarts Express** everywhere.